rust/core-lib/src/movement.rs

```diff
@@ -25,43 +29,158 @@ pub enum Movement {
     Left,
     /// Move to the right by one grapheme cluster.
     Right,
+    /// Move to the left by one word.
+    LeftWord,
+    /// Move to the right by one word.
+    RightWord,
+    /// Move to left end of visible line.
+    LeftOfLine,
+    /// Move to right end of visible line.
+    RightOfLine,
+    /// Move up one visible line.
+    Up,
+    /// Move down one visible line.
+    Down,
+    /// Move up one viewport height.
+    UpPage,
+    /// Move down one viewport height.
+    DownPage,
+    /// Move to the start of the text line.
+    StartOfParagraph,
+    /// Move to the end of the text line.
+    EndOfParagraph,
+    /// Move to the start of the document.
+    StartOfDocument,
+    /// Move to the end of the document
+    EndOfDocument,
+}
+
+/// Compute movement based on vertical motion by the given number of lines.
+///
+/// Note: in non-exceptional cases, this function preserves the `horiz`
+/// field of the selection region.
+fn vertical_motion(r: &SelRegion, view: &View, text: &Rope, line_delta: isize,
+    modify: bool) -> (usize, Option<HorizPos>)
+{
+    // The active point of the selection
+    let active = if modify {
+        r.end
+    } else if line_delta < 0 {
+        r.min()
+    } else {
+        r.max()
+    };
+    let col = if let Some(col) = r.horiz {
+        col
+    } else {
+        view.offset_to_line_col(text, active).1
+    };
+    // This code is quite careful to avoid integer overflow.
+    // TODO: write tests to verify
+    let line = view.line_of_offset(text, active);
+    if line_delta < 0 && (-line_delta as usize) > line {
+        return (0, Some(col));
+    }
+    let line = if line_delta < 0 {
+        line - (-line_delta as usize)
+    } else {
+        line.saturating_add(line_delta as usize)
+    };
+    let n_lines = view.line_of_offset(text, text.len());
+    if line > n_lines {
+        return (text.len(), Some(col));
+    }
+    let new_offset = view.line_col_to_offset(text, line, col);
+    (new_offset, Some(col))
 }
 
-/// Calculate a horizontal position in the view, based on the offset. Return
-/// value has the same type as `region_movement` for convenience.
-fn calc_horiz(view: &View, text: &Rope, offset: usize) -> (usize, Option<HorizPos>) {
-    let (_line, col) = view.offset_to_line_col(text, offset);
-    (offset, Some(col))
+/// Computes the actual desired amount of scrolling (generally slightly
+/// less than the height of the viewport, to allow overlap).
+fn scroll_height(view: &View) -> isize {
+    max(view.scroll_height() as isize - 2, 1)
```

Comment:
No, this is something else; it's the overlap (number of lines in common) on successive page scrolls. As slop is increased, this should stay the same. It should become a constant with a comment though.